Add draft project security threat-model document (#13293)
apache/cloudstack: draft-THREAT-MODEL.md

Proxy VM, virtual router); and orchestrates a fleet of **agents** running
on each hypervisor host. Authorization is RBAC + multi-tenant
domain/account/project hierarchy. The deployment shape is "operator-run
private/public cloud control plane", not a hosted-as-a-service appliance.

## §2 Scope and intended use

### Intended use

- A multi-tenant IaaS control plane deployed by an operator inside a
  controlled datacenter or cloud, exposing compute / storage / network
  orchestration to authenticated end users via a JSON API and a Vue.js Web
  UI, with separately authenticated administrators *(documented: `README.md`,
  `INSTALL.md`)*.
- Used both for service-provider public clouds and for on-premises private
  clouds; the trust model is the same in both *(documented: `README.md`)*.

### Deployment shape

CloudStack is **not** an in-process library, **not** a single-binary
appliance, and **not** a hosted SaaS. It is a distributed control plane:
one or more management-server instances (clustered behind a load balancer
in production), a MariaDB/MySQL database, one usage server, an optional
SecondaryStorageVM/ConsoleProxyVM/VirtualRouter set of system VMs, and a

> per-hypervisor-host `cloudstack-agent` (for KVM/Hyper-V/baremetal) or

Reviewer: ```suggestion
per-hypervisor-host `cloudstack-agent` (for KVM//baremetal) or
```

Author: Applied — line 107 now reads "per-hypervisor-host `cloudstack-agent` (for KVM/baremetal)". Thanks!

in-model only when it cleanly maps to one of them.

| # | Transition | Authentication | Authorization |
| --- | --- | --- | --- |
| B1 | API client → management server JSON API (`:8080`/`:8443`) | per-user API key + HMAC-SHA1 signature over query string, or session login + 2FA *(documented: `server/src/main/java/com/cloud/api/ApiServer.java` `verifyRequest`)*; signature version 3 has expiration enforcement *(documented: same file line ~1053)* | RBAC (dynamic-role-based / static-role-based / project-role-based) on the called API command name + domain/account ownership of named resources |
| B2 | Web UI → management server (`:8080`) | same as B1 plus session cookie | same as B1 |
| B3 | Browser → ConsoleProxyVM → hypervisor VNC socket | signed token issued by management server, embedded in URL; encrypted with `ConsoleProxyPasswordBasedEncryptor` *(documented: `server/src/main/java/com/cloud/servlet/ConsoleProxyServlet.java`, `ConsoleProxyPasswordBasedEncryptor.java`)* | implicit (signed-token possession) |
| B4 | Management server ↔ management server (cluster peers) | NIO + TLS, Root CA-issued certs *(documented: `framework/cluster/`, `framework/ca/`)* | peer-trust by valid cert |

> | B5 | Management server → `cloudstack-agent` (KVM/Hyper-V/baremetal) | NIO + TLS on `:8250`; agent uses X.509 client cert issued by Root CA on first connect; cert provisioning is the `SetupKeyStoreCommand` shape *(documented: `agent/src/main/java/com/cloud/agent/Agent.java` `setupAgentKeystore`, `framework/ca/.../CAService.java`, `plugins/ca/root-ca/.../RootCAProvider.java`)*; trust strictness governed by `ca.plugin.root.auth.strictness` (**default `false`** — see §5a) and `ca.plugin.root.allow.expired.cert` (**default `true`** — see §5a) | peer-trust by valid cert |

Reviewer 2: default value for `ca.plugin.root.auth.strictness` is true for newer setups but for older setups, it remains false even after upgraded.

Reviewer: ok, so if it is mentioned in the upgrade instructions it is not a concern.

Author: Folded — §5a/§11a note `ca.plugin.root.auth.strictness` default `true` (new) / `false` only on pre-Aug-2017 upgrade; treated as upgrade-instruction item, not a concern. Thanks!

  - **does** maintain MariaDB connections from the management server;
  - **does** issue X.509 certificates from its self-signed Root CA *(documented:
    `plugins/ca/root-ca/.../RootCAProvider.java`)*;
  - **does** spawn child processes from the agent (`Script` invocations
    against `/usr/share/cloudstack-common/scripts/`) *(documented:
    `agent/src/main/java/com/cloud/agent/Agent.java` `keystoreSetupSetupPath`,
    `keystoreCertImportScriptPath`)*;
  - **does** write logs under operator-configured locations;
  - **does** read a documented set of environment variables and the
    `db.properties` file at startup *(inferred — §14 Q16)*;
  - **does** install signal handlers / shutdown hooks only as
    Tomcat/Jetty servlet container default *(inferred — §14 Q16)*.

## §5a Build-time and configuration variants

CloudStack ships as a family of `cloudstack-management`, `cloudstack-agent`,
`cloudstack-usage`, `cloudstack-cli`, `cloudstack-ui` packages
*(documented: `debian/`, `packaging/`)*. A sizable number of runtime
configuration knobs materially change the security envelope. The
security-relevant subset:

| Knob | Default | Maintainer stance | Effect |
| --- | --- | --- | --- |

> | `ca.plugin.root.auth.strictness` | **`false`** *(documented: `RootCAProvider.java` line 132)* | **maintainer ruling required**: is the default a supported production posture or a dev-mode setting operators must flip per §10? *(inferred — §14 Q12)* | When `false`, the management server's `RootCACustomTrustManager` does **not** require a client certificate from a peer attempting to connect on `:8250` (agent port) or cluster ports. A peer without a cert is allowed in. |

Reviewer 2: Default is true. It's false only when upgrading from really old cloudstack versions.

Author: Folded — default `true`; `false` only when upgrading from very old versions (§5a). Thanks!

| `ca.plugin.root.allow.expired.cert` | **`true`** *(documented: `RootCAProvider.java` line 138)* | **maintainer ruling required** *(inferred — §14 Q12)* | When `true`, an expired client cert is accepted during SSL handshake. |
| `ca.plugin.root.issuer.dn` | `CN=ca.cloudstack.apache.org` *(documented: same file line 128)* | configured at first management-server boot | Subject DN of the auto-generated self-signed Root CA. |

> | `useforwardheader` (`use.forward.header`) | `false` *(inferred — §14 Q17)* | When `true`, the operator must restrict `proxy.forward.list` to the trusted reverse-proxy CIDR | When set, `ApiServlet.getClientAddress` honours `X-Forwarded-For` / configured headers *only* for source IPs in `proxy.forward.list` *(documented: `server/src/main/java/com/cloud/api/ApiServlet.java` lines 700–725)*. |

Reviewer 2: The name of the global setting is `proxy.header.verify`.

Author: Folded — corrected to `proxy.header.verify`. Thanks!

> | `proxy.forward.list` | unset *(inferred — §14 Q17)* | required when `useforwardheader=true` | CIDR list of trusted reverse proxies. |

Reviewer 2: `proxy.header.names` is list of names to check for allowed ipaddresses from a proxy set header.
`proxy.cidr` is a list of cidrs for which "proxy.header.names" are honoured if the "Remote_Addr" is in this list.

Author: Folded — §11a now states `proxy.header.names` is honoured only when `Remote_Addr` is in `proxy.cidr`. Thanks!

| `enable.2fa.for.users` / `enable.2fa.for.api` | per-domain toggle *(documented: `plugins/user-two-factor-authenticators/`)* | dev-test default off; production posture depends on PMC ruling *(inferred — §14 Q18)* | When on, users must complete static-pin or TOTP 2FA after login. |
| `security.encryption.key`, `security.encryption.iv` | auto-generated at first boot *(documented: `framework/security/.../KeysManager.java`)* | trusted secret | Base64-encoded JaSypt master key + IV used to encrypt application-level secrets in the DB. |

> | `auth.password.algorithm` (`hash.user.password`) | bcrypt / pbkdf2 / sha256salted *(documented: `plugins/user-authenticators/{pbkdf2,sha256salted}`)* | **maintainer ruling required**: which is the supported default for new deployments? `md5` and `plain-text` plugins still ship *(documented: `plugins/user-authenticators/{md5,plain-text}`)* — are these legacy-compat-only or in supported production? *(inferred — §14 Q19)* | governs how user passwords are stored |

Reviewer 2: No mention of either `auth.password.algorithm` or `hash.user.password` in code.

Order is defined by `user.password.encoders.order`. Default is `PBKDF2,SHA256SALT,MD5,LDAP,SAML2,PLAINTEXT`.  And to exclude `user.password.encoders.exclude` which has the default value - `MD5,LDAP,PLAINTEXT`.

Author: Folded — corrected to `user.password.encoders.order` (default PBKDF2,SHA256SALT,MD5,LDAP,SAML2,PLAINTEXT) + `.exclude` (MD5,LDAP,PLAINTEXT). Thanks!

> | `api.signature.version` | accepts both v1 and v3 *(documented: `ApiServer.java` line ~1053)* | v1 lacks an `expires` parameter; v3 requires expiration | A request with v3 + an expired `expires` is rejected; a v1 request without `expires` is accepted |

Reviewer 2: This isn't configurable and I didn't find any references for `api.signature.version` in the code.

Author: Folded — removed the `api.signature.version` reference (not configurable / not in code). Thanks!

> | `post.requests.and.timestamps.enforced` | per `isPostRequestsAndTimestampsEnforced` *(documented: `ApiServer.java` line ~1074)* | bounds `expires` to a maximum future offset | Prevents an attacker who steals a signed URL with a 10-year expiration from using it forever |

Reviewer 2: Global setting is `enforce.post.requests.and.timestamps` and not `post.requests.and.timestamps.enforced`

Author: Folded — corrected to `enforce.post.requests.and.timestamps`. Thanks!

| `integration.api.port` (`:8096`) | typically disabled *(inferred — §14 Q20)* | When non-zero, exposes an *unauthenticated* admin API for integration testing | An open integration port is a complete RBAC bypass on the management server |
| Hypervisor enablement (which `plugins/hypervisors/*` are installed and configured) | per zone | operator-driven | An unused hypervisor plugin still ships but is not connected to any host |

> | Hostname / SAN of management-server cert (`ca.plugin.root.management.cert.custom.san`) | unset *(inferred — §14 Q15)* | when set, included in the auto-generated cert SAN | governs which hostnames clients can use to reach the management server |

Reviewer 2: global setting is `ca.framework.cert.management.custom.san`

Author: Folded — corrected to `ca.framework.cert.management.custom.san`. Thanks!

| `OUT-OF-MODEL: equivalent-harm` | An actor already-authorized under the model can cause the same harm via a documented path (root admin doing root-admin things, RBAC-licensed user using their RBAC-licensed commands). | §3 items 4, 5 |
| `BY-DESIGN: property-disclaimed` | Concerns a §9 property the project explicitly does not provide (template sandboxing, side-channel resistance, hypervisor isolation, etc.). | §9 |
| `KNOWN-NON-FINDING` | Matches a §11a recurring false positive. | §11a |
| `MODEL-GAP` | Cannot be cleanly routed to any of the above — triggers §12 model revision. | §12 |

## §14 Open questions for the maintainers

Every *(inferred)* tag in the body maps to one of these. Proposed answers
are inline; please confirm, correct, or strike.

### Wave 1 — scope, intended use, the two big insecure defaults

**Q1.** The model assumes CloudStack is "a clustered distributed
control plane deployed inside an operator-controlled datacenter
network", not a single-host appliance or a hosted SaaS. Confirm? *(maps

> to §2)*

Reviewer: @potiuk , not sure if I should answer you but, the control plane could be clustered but for smaller clouds it can be a single instance of the management server. I am not sure if that is relevant for the question and I am not sure this is the proper way to answer the question.

Author: Folded into §2 (line 107) — single management-server instance or clustered, both supported. Thanks!

**Q2.** Are the SecondaryStorageVM, ConsoleProxyVM, and Virtual Router
treated as trusted-once-enrolled peers (proposed: **yes**, same shape as

> agents), or do they get their own trust tier? *(maps to §2, §4 B5)*

Reviewer: yes, for as far as they are pointing inwards. they do have for some purposed outside interfaces which are supposed to be safe.

Author: Folded — §2/§4 note the system-VM surfaces are inward-facing (with some intentionally-safe outside interfaces). Thanks!

**Q3.** Are external integrations (LDAP, SAML2 IdP, OAuth2 IdP, NSX
controller, Netscaler, Tungsten, S3-compatible storage, backup
providers) modeled as trusted control-plane peers (proposed: **yes**)? If
trusted, that licenses §3 item 2 and §11a trusted-input dispositions.

> *(maps to §2, §3, §11a)*

Reviewer: **yes**

Author: Confirmed — §14 Q3 resolved (external integrations = trusted control-plane peers). Thanks!

**Q4.** SecondaryStorageVM HTTP download surface — is the URL token
per-template ACL-checked, or is the SSVM URL itself a bearer credential

> that any holder can replay? *(maps to §6, §11a)*

Reviewer: I think the only mitigation to unauthorised use is timed availability of the download(-token) Not sure how to respond to this. We’ll have to make sure/test this. cc @vishesh92.
@potiuk, I wonder why code analysis did not discover this?

Reviewer 2: @DaanHoogland AFAIK, we create a symlink to generate a download link to apache http server to allow downloading of files. The symlink is in the UUID format to prevent enumeration attacks. And IIRC, the symlinks are removed after certain period.
There is no auth on the download links generated.

Author: Folded — §6/§11a: UUID symlinks, no auth on the link, removed after a window → `BY-DESIGN` (timed availability); §14 Q4 resolved. Noted the re-test-in-code point and the static-analysis question (a scan-agent note). Thanks!

**Q5.** Vendored upstream code under `systemvm/agent/noVNC/vendor/pako`
and bundled JaSypt / Bouncy Castle / JSch — is the policy "report
upstream; we pick up fixes on next sync" (proposed)? *(maps to §3 item 8,

> §11a)*

Reviewer: We have no procedure to implement this yet (dependabot is not producing viable PRs for us) I would prefer we do. Any ideas @vishesh92 ?

Reviewer 2: For noVNC, we are cloning the github.com/novnc/novnc and making some changes on top of it for cloudstack requirements.

Author: Folded — §3 item 8 / §11a: noVNC is a vendored fork (bugs upstream) and there's no automated update procedure today (the PMC would like one); §14 Q5 resolved. Thanks!

**Q6.** Is "an operator with `root` on a management-server host, the
JCEKS keystore + encryption keys, the Root CA private key, or MariaDB
credentials" out of scope (proposed: **yes**, `OUT-OF-MODEL:

> adversary-not-in-scope`)? *(maps to §3 item 1, §9)*

Reviewer: **yes**

Author: Confirmed — §14 Q6 resolved (operator-with-root = `OUT-OF-MODEL: adversary-not-in-scope`). Thanks!

**Q7.** Hypervisor bugs (libvirt / vSphere SDK / XenAPI / Hyper-V API /
KVM/QEMU itself) — out of scope, report upstream (proposed)? *(maps to

> §3 item 3)*

Reviewer: yes, should not lead to any changes but deleting this question, right @potiuk ?

Author: Confirmed — §14 Q7 resolved (hypervisor bugs out of scope, report upstream); question struck. Thanks!

### Wave 2 — the two big insecure defaults

**Q12.** **Highest-leverage question in the model.** Two Root-CA defaults:

- `ca.plugin.root.auth.strictness` = `false` *(documented:
  `RootCAProvider.java` line 132)* — the management server does **not**
  require a client cert from peers on `:8250` (agent port) and cluster
  ports by default. Is this **(a)** the supported production posture (so
  a report of "agent port accepts un-certed peer" is `VALID` against the
  default, meaning the default should be flipped), or **(b)** a dev/test
  convenience that operators are documented as required to flip per
  §10 (so the report is `OUT-OF-MODEL: non-default-build`)?
- `ca.plugin.root.allow.expired.cert` = `true` *(documented: same file
  line 138)* — expired client certs are accepted. Same question.

This single ruling reshapes §3 item 1, §5a, §7 (the un-certed peer
row), §8 P5, §10, §11 first/penultimate bullets, §11a first two
bullets, and §13. The text of §3 / §10 / §11a in this draft
**assumes the answer is (b)** — operator must flip both per §10.

> *(maps to §5a, §10, §11a, §13)*

Reviewer 2: For new setups, this is true by default. When upgrading from an older version (which didn't have this global setting which are versions released prior to Aug, 2017), this is false.

Ref.: https://github.com/apache/cloudstack/pull/2239

Author: Folded — §5a/§11a note new setups `true`; upgrades from pre-Aug-2017 `false` (ref #2239). Thanks!

DaanHoogland — `https://github.com/apache/cloudstack/pull/2239`)*:

- `ca.plugin.root.auth.strictness` defaults to **`true` on new setups** —
  the management server **does** require a Root-CA-signed client cert on
  `:8250` and the cluster ports. It is `false` **only** after upgrading
  from a version released before Aug 2017 that predates the setting; the
  upgrade instructions document turning it on, so a leftover `false` is an
  upgrade-hygiene gap, not a shipped insecure default.
- `ca.plugin.root.allow.expired.cert` defaults to `true` as an operational
  concession to cert-rotation lag.

This resolution reshaped §3 item 1, §5a, §7 (the un-certed peer row),
§8 P5, §9 false-friends, §10, §11, §11a, and §13. The earlier
"assumes operator must flip per §10" framing is withdrawn.

### Wave 3 — adjacent insecure defaults and admin-only surfaces

**Q8.** Is "a root admin with full RBAC role causes harm Y via a
documented path Z" out of scope (proposed: **yes**, `OUT-OF-MODEL:
equivalent-harm`)? In particular: `runCustomAction`, template upload,
plugin registration, global config change, system-VM patching, system-VM

> console access. *(maps to §3 item 4, §9)*

Reviewer 2: yes. A root admin will generally have direct access to most of these resources anyway.

**Q9.** Guest VM workloads — confirm that hypervisor-mediated side
channels and resource-exhaustion-within-allocation are out of scope, and
that the in-scope orchestration concerns are limited to "did CloudStack
place the VM in the right VLAN / apply the right security group / route

> the right IP" (proposed)? *(maps to §3 item 5, §7, §9)*

Reviewer 2: This sound right.
IMO, the only scenario where it would be a cloudstack problem will be if cloudstack is setting wrong/bad settings while launching the guest VM or some other action which results in the corresponding issues with hypervisor. In this scenario, CloudStack needs to ensure it's using correct/secure settings for the hypervisor.
@DaanHoogland What do you think?

**Q10.** Templates / ISOs / user-data — confirm that there is no
sandboxing of user-supplied OS images, and that user-data is intentionally
a code-execution channel into the guest (proposed)? *(maps to §3 item 6,

> §9)*

Reviewer 2: yes. that's correct. userdata is used by the end user to customize the guest OS.

**Q11.** Confirm the unsupported-component list: `tools/marvin/`,
`test/`, `developer/`, `quickcloud/`, `cloud-cli/`,
`tools/{devcloud4,devcloud-kvm,appliance,checkstyle,transifex,bugs-wiki,...}`,
`simulator` hypervisor plugin. Anything to add or remove? *(maps to §3

> item 7)*

Reviewer 2: @DaanHoogland do you think we need to include `simulator` & `tools/appliance`?

Reviewer: I think we need to *ex*clude them and make that explicit. Later we might want to create tooling with the express purpose of checking security but let’s leave oit out of scope for now.

**Q17.** Forward-header gating — the **setting names are confirmed**
*(maintainer: vishesh92)*: `proxy.header.verify` (the on/off gate),
`proxy.header.names` (header names to consult), and `proxy.cidr` (CIDRs of
the `Remote_Addr` values for which those headers are honoured). **Still
open:** confirm `proxy.header.verify` is **off by default** and that
`proxy.cidr` must be set for the headers to take effect. *(maps to §5a,

> §6, §10)*

Reviewer 2: `proxy.header.verify` is false by default. If the remote address is in the `proxy.cidr` list, CloudStack tries to get the correct value from the headers defined in `proxy.header.names`.

**Q18.** 2FA — proposed: off by default, operator turns it on per
domain / per user via `enable.2fa.*`. Confirm; and is "2FA disabled in
production" a §10 violation or a deployment choice? *(maps to §5a,

> §10)*

Reviewer 2: IMO, this is a deployment choice. The correct global settings for this are:
`enable.user.2fa` - default is `false`. Determines whether two factor authentication is enabled or not. This can also be configured at domain level.
`mandate.user.2fa` - default is `false`. Determines whether to make the two factor authentication mandatory or not. This setting is applicable only when enable.user.2fa is true. This can also be configured at domain level.

**Q19.** User-authenticator plugins — encoder selection is governed by
`user.password.encoders.order` (default
`PBKDF2,SHA256SALT,MD5,LDAP,SAML2,PLAINTEXT`) and
`user.password.encoders.exclude` (default `MD5,LDAP,PLAINTEXT`), so PBKDF2
is the effective hashing default and `MD5`/`PLAINTEXT` are retained only
for verifying legacy hashes *(maintainer: vishesh92)*. **Still open:**
confirm that a report against `md5`/`plain-text` being *used to hash new
passwords* in a greenfield install is `OUT-OF-MODEL: non-default-build`
(they are excluded by default), and confirm the supported greenfield

> encoder set. *(maps to §5a, §10, §11)*

Reviewer 2: a report against `md5`/`plain-text` being *used to hash new passwords* in a greenfield install is out of scope since it's excluded by default. The default encode set is `PBKDF2,SHA256SALT,MD5,LDAP,SAML2,PLAINTEXT` and we exclude `MD5,LDAP,PLAINTEXT"`. So the actual encode set will be - `PBKDF2,SHA256SALT,SAML2`.

**Q20.** Integration API port `:8096` — proposed: closed (port-zero) by
default in production packaging, open only when explicitly configured;
when open, it is unauthenticated by design. A report of "integration
port allows admin commands without auth" is `OUT-OF-MODEL:
non-default-build` *if* the operator opened it, else `VALID`. Confirm

> the default. *(maps to §5a, §10, §11a)*

Reviewer 2: The default should be 0 (disabled). But I need to confirm this.
@DaanHoogland any idea about this?

Reviewer: yes, it is set to 0 and only in test configurations it is set to 8096.

  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->

# Apache CloudStack Security Threat Model (draft)

> **Document scope and PMC structural decision.** The CloudStack PMC owns
> five repositories: `apache/cloudstack` (the management server, agent, and
> systemvm), plus four satellite clients — `apache/cloudstack-cloudmonkey`
> (CLI), `apache/cloudstack-go` (Go SDK), `apache/cloudstack-terraform-provider`,
> `apache/cloudstack-kubernetes-provider`. This document models
> `apache/cloudstack` as the canonical threat model; the four satellite
> models are short *deltas* that inherit §3 / §4 / §7 from this
> document and add only what each satellite uniquely introduces (`§4 B1`
> reachability, the credential file shape, the wrapper-of-SDK contract,

> > etc.). The deltas live at `/tmp/claude/cloudstack-<repo>-threat-model-draft.md`.

Reviewer 2: @potiuk as per my understanding we need to leave `/tmp` here for the jobs to work, right?

> The satellite clients' interfaces point **inward** at the management-server
> API; some satellites additionally expose **outward** interfaces that are
> designed to be safe to expose *(maintainer: DaanHoogland)*.
> An umbrella model was rejected because the satellites are uniformly thin
> "HMAC-SHA1-signing HTTP client" wrappers — a single document either
> drowns them in CloudStack-server content or, worse, drowns the
> CloudStack-server content in satellite caveats. Each satellite is small
> enough that a 1–2 page delta works.

## §1 Header

- **Project:** Apache CloudStack (`apache/cloudstack`) — IaaS orchestration
  platform. This document does **not** cover the four satellite repos, which
  carry their own delta models.
- **Commit:** `7308dad1` (HEAD of `main` at draft time).
- **Date:** 2026-05-29.

> - **Authors:** ASF Security team draft, awaiting CloudStack PMC review.

Reviewer: I don’t see how this is still valid, we can add maintainers but in my view the statsu already was “reviewed”/“approved”, and not “draft". guess it slipped through the cracks.

> - **Authors:** ASF Security team draft, awaiting CloudStack PMC review.

Reviewer: @potiuk , can you advice?

| **Hypervisor-managed guest VM (end-user workload)** | **untrusted** | A guest VM is an attacker's workload; the model defends against it. |
| **Reverse proxy / load balancer in front of management server** | trusted *(if `proxy.header.verify=true`)* | When the operator enables forward-header processing, only requests whose `Remote_Addr` ∈ `proxy.cidr` have their `proxy.header.names` header honoured *(documented: `server/src/main/java/com/cloud/api/ApiServlet.java` `getClientAddress`; setting names maintainer: vishesh92)*. |
| **Underlying storage (primary / secondary)** | trusted by virtue of operator-granted credentials | CloudStack reads/writes via NFS / RBD / iSCSI / S3 with operator-supplied credentials *(documented: primary/secondary storage plugins under `plugins/storage/`)*. |
| **External integrations (Tungsten, NSX, Netscaler, Palo Alto, …)** | trusted control-plane peers | Operator-configured; CloudStack assumes truthful responses *(inferred — §14 Q3)*. |

### Component-family table

| Family | Representative entry point | Touches outside the process? | In-model? |
| --- | --- | --- | --- |

> | Management server JSON and XML APIs | `client/.../ApiServlet`, HTTPS on `:8080` (admin), `:8080/client/api` (user), HTTPS on `:8443` integration port *(documented: `server/src/main/java/com/cloud/api/ApiServlet.java`, `client/`)* | network (TCP, optionally TLS) | **yes** |

Reviewer 2: this patch seems correct. @DaanHoogland please check this.

| Template / ISO upload | URL of remote source | **yes** within RBAC | upload-gated by `registerTemplate` RBAC; bytes are then served to hypervisors as image data |
| User-data / metadata service (`169.254.169.254` from inside guests) | guest-controlled bytes (the request) | **yes from the guest**, but the service is reached *from the guest* and serves only that guest's data | guest-VM-side isolation by virtual router |
| Hypervisor agent log / event stream | bytes from hypervisor | trusted operator surface | none — assumed truthful |
| LDAP / SAML / OAuth response (B6) | bytes from IdP | trusted | LDAP queries treat returned attributes as authoritative |
| Storage response (B8) | bytes / metadata from storage | trusted | bytes are object content; envelope is control-plane |

### Size / shape / rate

- CloudStack does not document a maximum signed-API request size; assumed
  to be servlet-container default (Jetty / Tomcat) *(inferred — §14 Q21)*.
- API rate limiting is per-account via the global config knobs `api.throttling.*`
  *(inferred — §14 Q22)*; an attacker with a valid API key can be rate-

>   limited at the application layer.

Reviewer 2: The params are there. I am not sure if this is working as expected or not.
@DaanHoogland any idea?

>   limited at the application layer.

Reviewer: @vishesh92 , as far as I understand the max size here is just for web-forms, not for API requests. The statement is about API requests so I think we should leave this one.

  their tags promoted from *(inferred)* to *(maintainer)*.

**About the project.** Apache CloudStack is an open-source Infrastructure-as-a-
Service (IaaS) orchestration platform *(documented: `README.md`,
`https://cloudstack.apache.org/`)*. It deploys and manages large fleets of
virtual machines across multiple hypervisors (KVM, VMware, XenServer/XCP-ng,
Hyper-V, baremetal-bridge, OVM) and over object/block/file storage
(NFS, Ceph/RBD, iSCSI, SMP, primary-storage plugins, S3-compatible secondary
storage). A central **management server** (Java/Tomcat-style servlets,
backed by MariaDB/MySQL) exposes a signed REST/JSON API to admins, end
users, and integrations; runs system VMs (Secondary Storage VM, Console
Proxy VM, virtual router); and orchestrates a fleet of **agents** running
on each hypervisor host. Authorization is RBAC + multi-tenant
domain/account/project hierarchy. The deployment shape is "operator-run
private/public cloud control plane", not a hosted-as-a-service appliance.

## §2 Scope and intended use

### Intended use

- A multi-tenant IaaS control plane deployed by an operator inside a
  controlled datacenter or cloud, exposing compute / storage / network

>   orchestration to authenticated end users via a JSON API and a Vue.js Web

Reviewer 3: ```suggestion
  orchestration to authenticated end users via APIs which return responses in JSON or XML and a Vue.js Web
```

API response can be in XML only, should we mention that?

| **Hypervisor-managed guest VM (end-user workload)** | **untrusted** | A guest VM is an attacker's workload; the model defends against it. |
| **Reverse proxy / load balancer in front of management server** | trusted *(if `proxy.header.verify=true`)* | When the operator enables forward-header processing, only requests whose `Remote_Addr` ∈ `proxy.cidr` have their `proxy.header.names` header honoured *(documented: `server/src/main/java/com/cloud/api/ApiServlet.java` `getClientAddress`; setting names maintainer: vishesh92)*. |
| **Underlying storage (primary / secondary)** | trusted by virtue of operator-granted credentials | CloudStack reads/writes via NFS / RBD / iSCSI / S3 with operator-supplied credentials *(documented: primary/secondary storage plugins under `plugins/storage/`)*. |
| **External integrations (Tungsten, NSX, Netscaler, Palo Alto, …)** | trusted control-plane peers | Operator-configured; CloudStack assumes truthful responses *(inferred — §14 Q3)*. |

### Component-family table

| Family | Representative entry point | Touches outside the process? | In-model? |
| --- | --- | --- | --- |

> | Management server JSON API | `client/.../ApiServlet`, HTTPS on `:8080` (admin), `:8080/client/api` (user), HTTPS on `:8443` integration port *(documented: `server/src/main/java/com/cloud/api/ApiServlet.java`, `client/`)* | network (TCP, optionally TLS) | **yes** |

Reviewer 3: Need to mention XML here?

Reviewer: ```suggestion
| Management server JSON and XML APIs | `client/.../ApiServlet`, HTTPS on `:8080` (admin), `:8080/client/api` (user), HTTPS on `:8443` integration port *(documented: `server/src/main/java/com/cloud/api/ApiServlet.java`, `client/`)* | network (TCP, optionally TLS) | **yes** |
```

Reviewer 3: @DaanHoogland there are few other places in document which may need similar change. I'm leaving this to your discretion.

Reviewer: I saw some mentions of JSON that do and others that do not also apply to XML. Maybe a generic remark that APIs may be used with XML as transport format is possible. I feel this can be added later.

   arbitrary network/storage plugins, and run `runCustomAction`-style
   commands. A new way for a root admin to do something they are already
   authorized to do is not a vulnerability *(maintainer: vishesh92 — §14 Q8)*. →
   `OUT-OF-MODEL: equivalent-harm`.
5. **A defender against a guest VM doing things the hypervisor allows it
   to do.** A guest VM consuming CPU, memory, or disk up to its allocated
   limit, sending arbitrary IP traffic within its assigned VLAN / VXLAN /
   security group, or exploiting another VM via the hypervisor's own
   shared surfaces (sidechannel, RowHammer, GPU passthrough leak) is out
   of model. CloudStack is responsible only for the orchestration that
   *places* the guest, not for hypervisor-level isolation *(maintainer:
   vishesh92 — §14 Q9; the in-model case is CloudStack applying
   wrong/insecure hypervisor settings — Daan to confirm boundary)*. → `OUT-OF-MODEL: adversary-not-in-scope` for the
   side-channel case, `BY-DESIGN: property-disclaimed` for the
   resource-limit case.
6. **A sandbox for templates, ISO images, or user-data scripts.** A
   user-uploaded template (via `registerTemplate`) is run by the

>    hypervisor with the privileges the offering grants. cloud-init /

Reviewer 3: I'm a bit unsure what we are trying to imply here. If we are talking about offerings from CloudStack, in my opinion they don't control privileges.

Reviewer: @shwstppr , you mean it should spell this:
```suggestion
6. **A sandbox for templates, ISO images, or user-data scripts.** A
   user-uploaded template (via `registerTemplate`) is run by the
   hypervisor with the privileges the system grants. cloud-init /
```
?

Reviewer 3: @DaanHoogland yes, makes more sense

  XenServer/XCP-ng, WinRM/Hyper-V API for Hyper-V).
- **What CloudStack does to its host** (negative claims, awaiting
  maintainer ratification):
  - **does** open listening sockets on documented ports
    (`:8080`/`:8443`/`:8250`/`:8096`/`:9090`/console-proxy ports) *(documented)*;
  - **does** maintain MariaDB connections from the management server;
  - **does** issue X.509 certificates from its self-signed Root CA *(documented:
    `plugins/ca/root-ca/.../RootCAProvider.java`)*;
  - **does** spawn child processes from the agent (`Script` invocations
    against `/usr/share/cloudstack-common/scripts/`) *(documented:
    `agent/src/main/java/com/cloud/agent/Agent.java` `keystoreSetupSetupPath`,
    `keystoreCertImportScriptPath`)*;
  - **does** write logs under operator-configured locations;
  - **does** read a documented set of environment variables and the
    `db.properties` file at startup *(inferred — §14 Q16)*;
  - **does** install signal handlers / shutdown hooks only as
    Tomcat/Jetty servlet container default *(inferred — §14 Q16)*.

## §5a Build-time and configuration variants

CloudStack ships as a family of `cloudstack-management`, `cloudstack-agent`,

> `cloudstack-usage`, `cloudstack-cli`, `cloudstack-ui` packages

Reviewer 3: No cloudstack-cli package. Better to mention cloudstack-agent?

Reviewer: good catch, there is no cli, packages created now are :
- cloudstack-agent
- cloudstack-baremetal-agent
- cloudstack-common
- cloudstack-integration-tests 
- cloudstack-management
- cloudstack-marvin
- cloudstack-mysql-ha
- cloudstack-ui
- cloudstack-usage

> `cloudstack-usage`, `cloudstack-cli`, `cloudstack-ui` packages

Reviewer: ```suggestion
CloudStack ships as a family of  packages: 'cloudstack-agent', 'cloudstack-baremetal-agent', 'cloudstack-common', 'cloudstack-integration-tests', 'cloudstack-management', 'cloudstack-marvin', 'cloudstack-mysql-ha', 'cloudstack-ui', 'cloudstack-usage'
```

Reviewer 3: Looks good @DaanHoogland

### Wave 4 — environment, distributed model, false-friends

**Q13.** Network-fabric assumptions — proposed: at least four logical
networks (management, public, guest, storage), with the management
network as the trusted control plane. Is that the canonical model, or
do you support more compressed topologies (single-fabric) in production? **RESOLVED** *(maintainer: DaanHoogland)* — there are four logical networks (management, public, guest, storage); each may have multiple instances across topologies (e.g. multiple zones) and may be combined within physical networks, but all four logical types must be present for a functional system. *(maps to §5, §10)*

**Q14.** Clock-skew assumption for signature v3 `expires` enforcement —
proposed: operator's responsibility to keep client + management-server clocks roughly in sync. Confirm. **RESOLVED** *(maintainer: DaanHoogland)* — confirmed; operator responsibility (PMC to add to the security model page). *(maps to §5)*

**Q15.** Confirm the filesystem-permissions inventory for sensitive
files: JCEKS keystore, Root CA private key, JaSypt key + IV,

> `db.properties`. Who owns them, what mode? **CLARIFIED** *(producer)* — not a CSV inventory of every file in a running system; only the four sensitive artifacts named here (JCEKS keystore, Root CA private key, JaSypt key + IV, `db.properties`), each with its owning UID and file mode. *(awaiting the per-file owner/mode values from the PMC.)* *(maps to §5, §10)*

Reviewer 2: The ownership of these files is the root user and the group is cloud. The permissions are read & write for root and read for cloud group.

**Q21.** API request size cap and cluster/agent RPC payload size cap —
are these explicitly bounded, or "whatever Jetty / NIO defaults give"? **RESOLVED** *(maintainer: DaanHoogland)* — the UI server sets an explicit cap, `org.apache.cloudstack.ServerDaemon.DEFAULT_REQUEST_CONTENT_SIZE = 1048576` (1 MiB); for other components the sizes are capped by the upstream components used. *(maps to §6, §9)*

**Q22.** `api.throttling.*` and per-account resource limits — proposed:
these are the entire DoS-protection surface, with no engine-level guard. Confirm. **RESOLVED** *(maintainer: DaanHoogland)* — confirmed; enforced at the API access check, and `api.throttling.enabled` is **`false` by default**. *(maps to §6, §9, §10)*

**Q23.** Decompression behaviour on uploaded QCOW2 / RAW / OVA — proposed:
no engine-side cap; per-account storage limits + hypervisor limits are the bound. Confirm. **RESOLVED** *(maintainer: DaanHoogland)* — correct. *(maps to §6, §9)*

**Q24.** Same-host non-`cloudstack` UID — proposed: game-over, no defence

> claimed. Confirm. *(maintainer: DaanHoogland notes there is a refusal to add a host with the same IP; whether that also includes a UID check is open — @vishesh92 to confirm. Disposition unchanged pending that.)* *(maps to §7, §9)*

Reviewer 2: @DaanHoogland the PR for this is still pending https://github.com/apache/cloudstack/pull/13182/
Currently when a host with the same IP is added again, we update the details for the existing host. But I don't think should be an issue since only a root admin/operator with access to the host credentials can do this.
I think adding a host directly with a configuration with fail due to missing key/certs required for connecting to the management server.

**Q27.** §8 P9 memory-safety — JVM-bounded; is the reachability
boundary correctly "in-model for the JSON API + B5 input; out-of-model
for native hypervisor SDK bugs that surface as `Throwable`"? **RESOLVED** *(maintainer: DaanHoogland)* — the reachability boundary is right, but **§8 P9 must not imply CloudStack is Java-only** — no implementation-language limitation is presumed (ocaml, python, bash run on hypervisors; go is used on the management server; the set may grow). The memory-safety claims hold for the JVM components only. *(reflected in §8 P9.)* *(maps to §8 P9, §9)*

**Q28.** §8 P10 listing-scope — confirm the §10 invariant "`list*`
responses are scoped to the principal's domain/account/project". And:
is information leak via error messages / async-job status / event log an in-model concern, or accepted? **RESOLVED** *(maintainer: DaanHoogland)* — in-model: regular system logs (e.g. log4j) are exempt, but other than those, information leaks (via error messages, async-job status, event log) are a concern. *(maps to §8 P10, §9, §11)*

**Q29.** Data-at-rest encryption — confirm CloudStack delegates entirely
to storage layer / hypervisor (LUKS, Ceph encryption, vSphere VM

> Encryption); no CloudStack-layer encryption of guest volumes. **RESOLVED** *(maintainer: DaanHoogland)* — correct (delegated to storage layer / hypervisor); *(vishesh92 to confirm)*. *(maps to §9)*

Reviewer 2: this is correct. delegated to storage layer / hypervisor.

for verifying legacy hashes *(maintainer: vishesh92)*. **RESOLVED** *(maintainer: vishesh92)* — a
report against `md5`/`plain-text` hashing *new* passwords in a greenfield
install is `OUT-OF-MODEL: non-default-build`: the default
`user.password.encoders.exclude` (`MD5,LDAP,PLAINTEXT`) removes them from
the effective set, so the supported greenfield encoders are
`PBKDF2,SHA256SALT,SAML2`. *(maps to §5a, §10, §11)*

**Q20.** Integration API port `:8096` — proposed: closed (port-zero) by
default in production packaging, open only when explicitly configured;
when open, it is unauthenticated by design. A report of "integration
port allows admin commands without auth" is `OUT-OF-MODEL:
non-default-build` *if* the operator opened it, else `VALID`. Confirm
the default. *(maps to §5a, §10, §11a)*

### Wave 4 — environment, distributed model, false-friends

**Q13.** Network-fabric assumptions — proposed: at least four logical
networks (management, public, guest, storage), with the management
network as the trusted control plane. Is that the canonical model, or
do you support more compressed topologies (single-fabric) in production?

> *(maps to §5, §10)*

Reviewer: There are four logical networks that can each have multiple instances for use in different topologies, e.g. multiple zones. They can be combined in physical networks. All four types of the logical networks must be present for a functional system.

**Q14.** Clock-skew assumption for signature v3 `expires` enforcement —
proposed: operator's responsibility to keep client + management-server

> clocks roughly in sync. Confirm. *(maps to §5)*

Reviewer: yes, (reminder @vishesh92 , this might be one we need to add to the security model page: https://cloudstack.apache.org/security)

**Q15.** Confirm the filesystem-permissions inventory for sensitive
files: JCEKS keystore, Root CA private key, JaSypt key + IV,

> `db.properties`. Who owns them, what mode? *(maps to §5, §10)*

Reviewer: @potiuk , do you expect a csv type list of all files in a functioning system?

**Q16.** Confirm the "what CloudStack does not do to its host" inventory
in §5: no child processes besides agent `Script` invocations / system
VM provisioning; signal-handlers via servlet container default;
environment-variable consumption confined to documented set. Anything to

> add? *(maps to §5)*

Reviewer: confirmed

**Q21.** API request size cap and cluster/agent RPC payload size cap —
are these explicitly bounded, or "whatever Jetty / NIO defaults give"?

> *(maps to §6, §9)*

Reviewer: the UI-server has an explicit request size: in org.apache.cloudstack.ServerDeamon; DEFAULT_REQUEST_CONTENT_SIZE = 1048576. For other components the sizes are capped by the upstream components used.

**Q22.** `api.throttling.*` and per-account resource limits — proposed:
these are the entire DoS-protection surface, with no engine-level

> guard. Confirm. *(maps to §6, §9, §10)*

Reviewer: confirmed, this is processed at API access check. (default of api.throttling.enabled == false!!)

**Q23.** Decompression behaviour on uploaded QCOW2 / RAW / OVA — proposed:
no engine-side cap; per-account storage limits + hypervisor limits are

> the bound. Confirm. *(maps to §6, §9)*

Reviewer: correct

**Q24.** Same-host non-`cloudstack` UID — proposed: game-over, no defence

> claimed. Confirm. *(maps to §7, §9)*

Reviewer: @vishesh92 , I think there is a refusal to add a host with the same IP, does this include a UID check as well (or should it)?

**Q25.** Side-channel observers (cache, branch, hypervisor-shared) — out

> of scope (proposed). *(maps to §7, §9)*

Reviewer: @potiuk I agree with cache and hypervisor-shared, (if I understand them correctly) but I do not understand “branch” in this context. Can you explain?

**Q26.** Byzantine-internal-peer threshold — confirm CloudStack makes no
BFT claim, so any compromised cluster peer or agent with a valid

> Root-CA-issued cert is unbounded (proposed). *(maps to §7, §9)*

Reviewer: Agreed. @vishesh92 we might want to add some issues/feature proposals in this area. This will only work in larger clusters, not in single or dual machine clusters (if I understand the byzantine model correctly).

**Q27.** §8 P9 memory-safety — JVM-bounded; is the reachability
boundary correctly "in-model for the JSON API + B5 input; out-of-model
for native hypervisor SDK bugs that surface as `Throwable`"? *(maps to

> §8 P9, §9)*

Reviewer: §8 P9 says that "CloudStack's own server-side code is Java”, implying it is only java. This is not correct. No limitation on implementation languages is presumed. Claims about the JVM are correct.

For instance ocaml and python code can run on hypervisors, as well as bash, and go is used on the management server. This list may be not complete now or in the future.

**Q28.** §8 P10 listing-scope — confirm the §10 invariant "`list*`
responses are scoped to the principal's domain/account/project". And:
is information leak via error messages / async-job status / event log

> an in-model concern, or accepted? *(maps to §8 P10, §9, §11)*

Reviewer: Regular system logs (log4j for instance) are exempt. Other than these, all information leaks are a concern.

**Q29.** Data-at-rest encryption — confirm CloudStack delegates entirely
to storage layer / hypervisor (LUKS, Ceph encryption, vSphere VM
Encryption); no CloudStack-layer encryption of guest volumes. *(maps to

> §9)*

Reviewer: correct (@vishesh92 , please confirm)

**Q30.** Constant-time comparison — confirm that *only* the API
signature path uses `ConstantTimeComparator`. Login password compare,
session cookie compare, console-token compare — none documented

> constant-time. Is that intentional? *(maps to §8, §9)*

Reviewer: I do not understand; Is the fact that it is not documented intentional?
If this is indeed the question, than yes, this is a lack of feature. (pretty sure I am missing the point, cc @vishesh92 )

**Q31.** Time-of-check-to-time-of-use between RBAC check at API entry
and orchestration on agent fleet — confirm mid-job RBAC revocation is

> **not** retroactively enforced (proposed). *(maps to §9)*

Reviewer: agreed/confirmed

**Q32.** TLS posture on `:8080` vs `:8443` — confirm production deploys
behind TLS on `:8443` or behind a TLS-terminating reverse proxy; a bare

> `:8080` HTTP API is dev-only. *(maps to §5a, §10)*

Reviewer: confirmed

**Q33.** `security.encryption.key` reuse across environments — confirm
that reusing the JaSypt key + IV across staging and production is a

> documented misuse. *(maps to §11)*

Reviewer: indeed

### Wave 5 — meta

**Q34.** Should this document live at `docs/threat-model.md` in
`apache/cloudstack`, or as a page on `cloudstack.apache.org/security/`?

> Or both, with one canonical and the other linked? *(meta)*

Reviewer: in my not so humble opinion: cloudstack.apache.org/security should contain an excerpt of, and a link to threat-model.md, the later being the source of truth. @vishesh92 ?

Reviewer 2: this sounds good. If required, we will change it later.

**Q35.** Is there an existing CloudStack threat-model document
(Confluence, internal, or a `[SECURITY]`-tagged dev@ thread) that this
should reconcile against rather than supersede? *(meta — §3.1a of the

> rubric)*

Reviewer: cloudstack.apache.org/security/ is the only security model at this moment and this should enhance this by providing it with a source of truth.

**Q36.** What kind of change should trigger a revision (proposed list in

> §12 — confirm or correct)? *(meta, §12)*

Reviewer: One that I would add is a change in the extension mechanisms implemented by CloudStack.

**Q37.** §11a is the highest-leverage section for the scan agent's
suppression list. The current draft has 15 patterns; could the PMC
populate §11a from recurring "not a vuln" closures on the
`security@apache.org` ↔ CloudStack triage queue and on
`https://cloudstack.apache.org/security.html`? Concrete asks: 3–5
patterns the PMC sees recur in inbound reports (e.g. "SSL bare on
`:8080` in a dev cluster", "agent port open without strictness flipped",
"`md5` authenticator left enabled after upgrade", "console URL appears
in support ticket"). *(meta — §11a)*

**Q38.** Confirm the structural decision to keep the four satellite repos
as separate delta models (`cloudstack-go-threat-model-draft.md`,
`cloudstack-cloudmonkey-threat-model-draft.md`,
`cloudstack-terraform-provider-threat-model-draft.md`,
`cloudstack-kubernetes-provider-threat-model-draft.md`) inheriting §3

> / §4 / §7 from this document. *(meta, §3 item 9)*

Reviewer: confirmed, these are not the system core. They can not be used without the core but the core can be used without them. There is in fact an added hierarchy to the repos in that cloudstack-go is a dependency to the other three.
